### PR TITLE
RFC: Restrict reflection queries to the requested schema

### PR DESCRIPTION
**Pull request by atrigent**
What do you guys think of these changes? I don't know if they're appropriate for upstream, but we may use them for our own purposes.

## Todos
- [ ] MIT compatible
- [ ] Tests
- [ ] Documentation
- [ ] Updated CHANGES.rst

### CODE REVIEW

**Comment by atrigent** [comment]
What would be your preferred way for me to reduce the width of this line?

```
        all_constraints = self._get_schema_constraint_info(connection, schema,
                                                           info_cache=info_cache)
```

File: sqlalchemy_redshift/dialect.py [hunk 1]
```diff
@@ -547,12 +543,11 @@ def create_connect_args(self, *args, **kwargs):
         return cargs, default_args
 
     def _get_table_or_view_names(self, relkind, connection, schema=None, **kw):
-        default_schema = inspect(connection).default_schema_name
         if not schema:
-            schema = default_schema
+            schema = self.default_schema_name
```
Comment (jklukas) [inline sqlalchemy_redshift/dialect.py]: Is self.default_schema_name provided by a superclass here?

**Comment by jklukas** [comment]
> What would be your preferred way for me to reduce the width of this line?

I'd recommend:
```
        all_constraints = self._get_schema_constraint_info(
            connection, schema, info_cache=info_cache)
```

File: sqlalchemy_redshift/dialect.py (hunk 1 above)
Comment (atrigent) [inline sqlalchemy_redshift/dialect.py]: Yes. See here: http://docs.sqlalchemy.org/en/latest/core/internals.html#sqlalchemy.engine.interfaces.Dialect. And the psycopg2 dialect which this dialect inherits from of course ultimately inherits from Dialect and implements this property.

Comment (jklukas) [inline sqlalchemy_redshift/dialect.py]: Using this attribute definitely cleans up the code compared to passing a connection around and calling inspect, which is very nice.

**Comment by jklukas** [comment]
The changes you have here are more straightforward than I had imaged when discussing in #104, so I don't see any downside to this approach compared to what we were doing previously. I'm in support of this change.

**Comment by jklukas** [comment]
I can't remember if there's a way for us to make TravisCI run the tests against Redshift for PRs from forks. @graingert - do we have a protocol for that?

**Comment by jklukas** [comment]
I'm going to try pulling the fork locally to run the full test suite.

**Comment by jklukas** [comment]
I ran this locally to get tests against Redshift. Everything passes, so +1 from me.

**Comment by jklukas** [comment]
I should qualify the +1 with the expectation that we add a line to CHANGES.rst explaining the change. @atrigent - can you add that in this PR?

**Comment by atrigent** [comment]
It occurred to me that the `RelationKey` stuff can probably go away now as well, since you'll only ever be dealing with things within a schema. However, I'm unsure what the purpose of the `unquoted` method is, so I'm hesitant to remove it. I also noticed that the `unquoted` method is probably wrong anyway, since it returns a string rather than another `RelationKey`...

**Comment by atrigent** [comment]
Unfortunately we just saw the offending error a few more times, from the `_get_schema_column_info` query. I looked at the explain output for that query, and noticed this:

```
->  LD Seq Scan on pg_class c  (cost=0.00..23.10 rows=163 width=136)
      Filter: pg_table_is_visible(oid)
```

The query isn't calling that function directly, so I believe the source of this function call is the use of the `pg_table_def` view, which is defined like:

```
 SELECT n.nspname AS schemaname, c.relname AS tablename, a.attname AS "column", format_type(a.atttypid, a.atttypmod) AS "type", format_encoding(a.attencodingtype::integer) AS "encoding", a.attisdistkey AS "distkey", a.attsortkeyord AS "sortkey", a.attnotnull AS "notnull"
   FROM pg_namespace n, pg_class c, pg_attribute a
  WHERE n.oid = c.relnamespace AND c.oid = a.attrelid AND a.attnum > 0 AND NOT a.attisdropped AND pg_table_is_visible(c.oid)
  ORDER BY n.nspname, c.relname, a.attnum;
```

I'm not sure I see why this view is being used. You're already joining with `pg_attribute` as well, and the only thing you take from `pg_table_def` is `column`, which you could equally well get from `pg_attribute.attname`. Is it for the `a.attnum > 0 AND NOT a.attisdropped` stuff? I'd be a shame to have to duplicate that in the query, but that might be what's necessary to fix this for us. On the plus side, removing use of `pg_table_def` would remove the need to monkey with the search path.

**Comment by atrigent** [comment]
Sorry, I just realized you are taking some other columns from `pg_table_def`. I wish there was just a version of this that didn't call `pg_table_is_visible`...

**Comment by atrigent** [comment]
@jklukas Can you comment on this comment of mine: https://github.com/sqlalchemy-redshift/sqlalchemy-redshift/pull/105#issuecomment-259848263 ?

**Comment by jklukas** [comment]
The `unquoted` method is intended to remove double quotes around qualified names if they exist, but was a bit of a hack. If tests pass with that removed, then we're probably fine. I don't remember the details of where that was important.

**Comment by jklukas** [comment]
Addressed in https://github.com/sqlalchemy-redshift/sqlalchemy-redshift/pull/223

**Pull request closed without merging**